### PR TITLE
Classic McEliece

### LINKED ISSUE #121: Add Round 2 KEM Classic-McEliece

### PR DESCRIPTION
**Pull request by thomwiggers**
* [x] Duplicate consistency files
* [x] Other implementations
  * [x] `vec`
  * [x] `sse`
  * [x] `avx2`

* [x] Update `mceliece*f/vec` from SUPERCOP-20191221

Closes #121 

#### Manually checked properties

* [x] No stringification macros
* [ ] Output-parameter pointers in functions are on the left
* [ ] Negative return values on failure of API functions (within restrictions of FO transform).
* [ ] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [ ] All integer types are of fixed size, using `stdint.h` types (including `uint8_t` instead of `unsigned char`)
  * [ ] Integers used for indexing are of size `size_t`

### CODE REVIEW

**Comment by thomwiggers** [comment]
The reference implementation assumes it's running on a little-endian machine. I've sent an email to DJB, to discuss how to get this resolved.

File: crypto_kem/mceliece348864/clean/pk_gen.c [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#include "controlbits.h"
+#include "benes.h"
+#include "params.h"
+#include "pk_gen.h"
+#include "root.h"
+#include "util.h"
+
+/* input: secret key sk */
+/* output: public key pk */
+int PQCLEAN_MCELIECE348864_CLEAN_pk_gen(uint8_t *pk, uint8_t *sk, uint32_t *perm) {
```
Comment (thomwiggers) [inline crypto_kem/mceliece348864/clean/pk_gen.c]:
```suggestion
int PQCLEAN_MCELIECE348864_CLEAN_pk_gen(uint8_t *pk, uint32_t *perm, const uint8_t *sk) {
```

File: crypto_kem/mceliece348864/clean/pk_gen.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#ifndef PK_GEN_H
+#define PK_GEN_H
+/*
+  This file is for public-key generation
+*/
+
+
+#include "gf.h"
+
+int PQCLEAN_MCELIECE348864_CLEAN_pk_gen(unsigned char * /*pk*/, unsigned char * /*sk*/, uint32_t * /*perm*/);
```
Comment (thomwiggers) [inline crypto_kem/mceliece348864/clean/pk_gen.h]:
```suggestion
int PQCLEAN_MCELIECE348864_CLEAN_pk_gen(uint8_t * /*pk*/, uint32_t * /*perm*/, uint8_t *sk);
```

**Comment by thomwiggers** [comment]
They're not as clean as they could probably be (many opportunities for better types, still some `static inline` functions in `sse` and `avx` implementations, function argument ordering, ...), but I think they're in a pretty good shape now.

**Comment by thomwiggers** [comment]
The parent repo of this pull request is https://github.com/thomwiggers/mceliece-clean/ but it's not much clearer than this repo, unfortunately – you will need to use ``rmlint`` to figure out which files are duplicate.

File: crypto_kem/mceliece348864/avx/transpose.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+
+#include "vec256.h"
+
+#include <stdint.h>
+
+extern void PQCLEAN_MCELIECE348864_AVX_transpose_64x64_asm(uint64_t *);
+
+static inline void PQCLEAN_MCELIECE348864_AVX_transpose_64x64(uint64_t *in) {
+    PQCLEAN_MCELIECE348864_AVX_transpose_64x64_asm(in);
```
Comment (mkannwischer) [inline crypto_kem/mceliece348864/avx/transpose.h]: can you move that to a c file?

File: crypto_kem/mceliece348864/clean/api.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    unsigned char *key,
+    const unsigned char *c,
+    const unsigned char *sk
+);
+
+int PQCLEAN_MCELIECE348864_CLEAN_crypto_kem_keypair
+(
+    unsigned char *pk,
+    unsigned char *sk
+);
```
Comment (mkannwischer) [inline crypto_kem/mceliece348864/clean/api.h]: those should be fixed size integers

File: crypto_kem/mceliece348864/avx/int32_sort.h [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#ifndef int32_sort_h
```
Comment (mkannwischer) [inline crypto_kem/mceliece348864/avx/int32_sort.h]: why do  some of the header guards have namespacing and others don't?

File: crypto_kem/mceliece348864/avx/operations.h [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#ifndef OPERATIONS_H
+#define OPERATIONS_H
+
+#include "api.h"
+
+#endif
+
```
Comment (mkannwischer) [inline crypto_kem/mceliece348864/avx/operations.h]: this file is pretty useless, remove?

File: crypto_kem/mceliece348864/avx/vec.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+static inline void vec_add(uint64_t *h, const uint64_t *f, const uint64_t *g) {
+    int b;
+
+    for (b = 0; b < GFBITS; b++) {
+        h[b] = f[b] ^ g[b];
+    }
+}
+
```
Comment (mkannwischer) [inline crypto_kem/mceliece348864/avx/vec.h]: can you move those to a c file?

File: crypto_kem/mceliece348864/clean/benes.c [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            data[j + s] ^= d;
+        }
+    }
+}
+
+/* input: r, sequence of bits to be permuted */
+/*        bits, condition bits of the Benes network */
+/*        rev, 0 for normal application; !0 for inverse */
+/* output: r, permuted bits */
+void PQCLEAN_MCELIECE348864_CLEAN_apply_benes(unsigned char *r, const unsigned char *bits, int rev) {
```
Comment (mkannwischer) [inline crypto_kem/mceliece348864/clean/benes.c]: fixed size integers?

File: crypto_kem/mceliece348864/avx/int32_sort.h (hunk 5 above)
Comment (thomwiggers) [reply crypto_kem/mceliece348864/avx/int32_sort.h]: Consequence of this being a manual effort – we only really require this for ``api.h`` files. I'll see if I can craft a regex.

Comment (thomwiggers) [reply crypto_kem/mceliece348864/avx/int32_sort.h]:
```sh
 perl -p -e 's/^(#(ifndef|define)) +(MC_)?(([a-zA-Z0-9_]|_)+_[hH])$/\1 MC_\U\4/' -i $(find schemes -iname '*.h')
```

File: crypto_kem/mceliece348864/clean/benes.c (hunk 8 above)
Comment (thomwiggers) [reply crypto_kem/mceliece348864/clean/benes.c]: We don't require that outside of `api.h`.

**Pull request merged**